pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go

```diff
@@ -183,8 +183,8 @@ func (dswp *desiredStateOfWorldPopulator) isPodTerminated(pod *v1.Pod) bool {
 // exist but should
 func (dswp *desiredStateOfWorldPopulator) findAndAddNewPods() {
 	for _, pod := range dswp.podManager.GetPods() {
-		if dswp.isPodTerminated(pod) {
-			// Do not (re)add volumes for terminated pods
+		if dswp.isPodTerminated(pod) || util.IsTerminating(pod) {
```

Comment:
I discussed this a bit with @jingxu97 and agree that we need to be really careful here.

When the Pod has a deletion timestamp, that triggers the containers to be killed. Once all the containers are killed, then the Pod phase is set to terminated, and only then it is safe to start unmounting the volumes.  So by adding an IsTerminating check here, we could potentially introduce some condition where volumes start to be unmounted before the containers have been terminated.

We think to correctly fix this issue, we need to figure out why exactly the Pod got added back when the phase should have already been Terminated (because we unmounted the volumes)

Reply:
@jingxu97 was able to find the problem.  The issue is pods are removed from dsw based on podStatusManager, but pods are added to dsw based on podManager.  podStatus cache is updated before making an API call, and podManager cache is updated after the API call, so those 2 caches can have out of sync states.  We discussed a bit with @yujuhong and think the best solution is to only use podManager to determine if pod is terminated or not, so we only operate based off of one cache instead of two.  The downside of this is that pod deletion latency may increase because we need to wait until the API update gets reflected in podManager before starting to unmount volumes.  And if API server is unavailable, then we won't unmount volumes either.

@dashpole was the last person to modify the [terminated check](https://github.com/kubernetes/kubernetes/blame/master/pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go#L175). Do you have any thoughts on this?

Reply:
That is one possible solution.  Could we instead have them both use the status manager?  That would eliminate the need for the round trip.

Reply:
The problem with using pod status manager is that on a kubelet restart, pod status manager doesn't get populated from the apiserver during init.  So it's possible the pod may not exist in pod status manager for a while until the syncPod happens. But because volume reconciling happens asynchronously, the temporarily absent pod could cause volume manager to unmount the volumes.

Reply:
Other components have a similar problem.  What we do for them is to use the status manager, if available, and fall back to the pod manager, if the status manager does not have it.

The pod passed in here: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_pods.go#L852 is from the pod manager

Reply:
Yes this is the problem.  The pod was terminated in the status manager, so we removed the volume from dsw and started and finished volume unmounting.  However the pod was still shown as terminating in the podManager cache, so we added the volume back into dsw and remounted the volumes again.

Reply:
If fsgroup check is not important, we can disable the tests temporarily, but issue exists anyway. If volume may be mounted again even if it has be detached from this node, it violates attachdetach controller, there may be other problems (e.g. volume plugins may rely on attachdetach controller to make sure RWX/ROX PV will only be mounted on one node at a time). I think we should make sure if volumes are detached from node should not be remounted on this node again.

> The issue is pods are removed from dsw based on podStatusManager, but pods are added to dsw based on podManager. podStatus cache is updated before making an API call, and podManager cache is updated after the API call, so those 2 caches can have out of sync states.

I'm not familiar with this part of code, is it possible to sync cache between statusManager and  podManager?